### ISSUE TITLE
When pixels are moved, they get the layer opacity remultiplied into them, and their opacity is used to blend them into the target location in the same layer

### ISSUE THREAD

**Issue by logiclrd** [PintaProject/Pinta#819]
**Description**
If you select a set of pixels in a layer, the Move Selected Pixels tool can be used to change their position within the layer. When the move operation completes and the pixels are deselected, they are recomposited into the layer. But, if the layer's opacity is not 100%, the pixels which were moved end up having the layer opacity remultiplied against them as part of the commit operation.

Similarly, if the pixels themselves are not 100% opaque, that opacity is used to blend them back into the layer they're being moved within. Committing a move should simply overwrite the pixels -- including the opacity, so that they bring the opacity they had where they were moved from but don't get blended back into their layer.

**To Reproduce**
1. Create an image.

![image](https://github.com/PintaProject/Pinta/assets/2349379/33deae90-07ec-4fd6-a329-214b752b9e57)

2. Paste another image in as a new layer.

![image](https://github.com/PintaProject/Pinta/assets/2349379/502a71f7-a443-4b1b-a83e-dd078f23df78)

3. Set the layer opacity to 50%.

![image](https://github.com/PintaProject/Pinta/assets/2349379/138bb444-387d-4fba-8b75-5bd9d94464bf)

4. Select a region of pixels in the layer.

![image](https://github.com/PintaProject/Pinta/assets/2349379/79da524c-4819-43ad-a4a2-dafa1c3b7267)

5. Using the Move Selected Pixels tool, move the pixels to another location.

![image](https://github.com/PintaProject/Pinta/assets/2349379/4d9f449f-d8f0-4493-b5ba-4fa403d7f0bd)

The problem is already apparent here, because the moved region is transparent and the part of the layer behind where the pixels have been moved to can be seen through the selection.

6. Deselect the pixels to commit the move operation.

![image](https://github.com/PintaProject/Pinta/assets/2349379/f1ab8674-7d48-4af0-af38-9528d6c501cb)

The pixels have been cut from the layer and then pasted into the new location, but the layer's opacity was included in the cut and applied in the paste. The destination for the paste, though, is the layer which has that opacity as a persistent property as well, so now the opacity is applied to those pixels twice. In addition, the newly-pasted pixels show through what was beneath them in the layer. They should have simply replaced the pixels in the layer. With respect to the paste operation, they should be 100% opaque.

**Version**
Operating system: Ubuntu 24.04
Pinta version: 2.2 built from 60cabb813f8397fc9a2b4ab946ce373259d550db

**Comment by cameronwhite** [comment]
I think you might want to use the `Cairo.Operator.Source` blend mode when flattening the selection back on to the layer? That should just replace the pixels rather than blending them with what's underneath

(https://www.cairographics.org/operators/)

**Comment by logiclrd** [comment]
That seems reasonable, though fixing this is going to be more than a one-line fix I suspect. During the drag operation, if I'm understanding correctly, the selection layer is modelled as just another layer, composited on top of the layers that are part of the image. It simply can't work that way with respect to alpha. You would need to have some sort of hierarchical layer system.

Let's say you have two layers, layer 1 background and layer 2 with some shape with partial translucency on it. If you select some of the pixels in layer 2 and start moving them, you now need to have a separate stack of layers _within_ layer 2 that represent the selection overlaid onto the remainder of layer 2's pixels. This operation preserves the full RGBA for every pixel but doesn't process alpha at all, simply replaces the alpha pixels. Then, the result of that compositing is the pixels used for layer 2 in the main stack of layers, and there it does get composited with full alpha processing. With this scheme, layer 1 shows through where layer 2 is translucent, but within layer 2's processing, there is no alpha interaction at all with the selection that's being moved -- but that selection _still keeps its alpha channel data_. Does that make sense?

**Comment by cameronwhite** [comment]
Yeah I think that makes sense to me. Probably a good first step would be to fix the final update of the layer so at least the final result is as expected, even if the preview is a bit off

**Comment by logiclrd** [comment]
Could you point me at where that takes place? :-)

**Comment by cameronwhite** [comment]
> Could you point me at where that takes place? :-)

I think the final write back will happen in `Document.FinishSelection ()`

**Comment by logiclrd** [comment]
Thanks, I'll take a look.

**Comment by cameronwhite** [comment]
This got auto-closed by the PR, but I think we still want to keep this open for getting the correct blending before the selection is committed

**Comment by logiclrd** [comment]
Makes sense :-)
